### ISSUE TITLE
broken for Nook?

### ISSUE THREAD

**Issue by unhandyandy** [noDRM/DeDRM_tools#7]
Nook deDRM has just stopped working for me, after years of success.  Any idea what's changed?

**Comment by ElleKayEm** [comment]
Your key probably changed.  Here's how to extract it now: https://www.aricrenzo.com/2019-12-13-Liberate-Your-Nook-Ebooks/

**Comment by unhandyandy** [comment]
I tried that method but the cc_hash_data table has no rows.  I found another cc_hash_data table in devices.db, which had two rows, but both were 60 characters long and neither worked.

**Comment by ElleKayEm** [comment]
Some people who had that issue had better luck with a different virtual Android device and or Nook app.  You might find some useful info in more recent comments in this thread:  https://github.com/apprenticeharper/DeDRM_tools/issues/814

**Comment by unhandyandy** [comment]
Yeah, I had seen that thread, it didn't work on my rooted Samsung tablet.  Are you saying I should try a virtual device instead?  That was never necessary before.

And are saying there are "different ... Nook app[s]" I could try?  Do you mean an earlier version of the Nook app?

**Comment by ElleKayEm** [comment]
Yes, that's what I meant.  You might also try logging out and back into the Nook app on your tablet.  Seems like someone mentioned that working for them.

**Comment by momijizukamori** [comment]
Found this issue while doing some hunting of my own, so some more info! I also experienced the same issue trying to find the key hash in the version of the Nook app directly from the Google Play store (v5.5.1.12) - no rows in the cc_hash_data table of cchashdata.db, 60 character hash in the cc_hash_data table in devicemanager.db which doesn't work to decrypt the epubs. 

HOWEVER. I decided to try an older version of the app, from the site listed in the linked blog post (v5.0.5.35, because the blog post discusses a 5.0 version working. The cchashdata.db file generated after signing in on THAT version does have a row with a key in it, which is 28 characters long. And, very interestingly, importing that key in the Calibre plugin works on the epubs I downloaded and copies over on the higher version. So I'm guessing either the 60-character key is a different representation of the same data in the 28-character key (which the plugin isn't equipped to handle yet), or there's a 28-character key hidden somewhere else in the files of the newer version that I haven't been able to find yet (though I did poke around the other dbs some and found nothing).

**Comment by noDRM** [comment]
The actual key for the encryption itself (the one that needs to be imported into the plugin) will always be 28 characters. There's no way for Nook to change that without breaking compatibility for all existing readers. That's because the key is the base64 representation of a SHA1 hash (using Adobe's PassHash implementation). If the key wasn't 28 characters / 20 bytes, it would no longer be the passHash algorithm. This means that there must be a 28-character string hidden somewhere else. 

Of course the key itself, the data encoded in this 28-character base64 string can be changed by Nook. 

In case you didn't see it yet, fesiwi created a Python script that can export this key from the Windows Store Nook app: https://github.com/noDRM/DeDRM_tools/discussions/9 
I plan to integrate this script into the plugin as soon as possible.

**Comment by unhandyandy** [comment]
Can fesiwi's script be adapted to Android?

**Comment by noDRM** [comment]
Yes, it can, and I've just done that :P

Only issue would be extracting the necessary data, as it looks like the Nook app forbids backups by default. 

Yes, it can be adapted to Android, and I've just done that for a quick test (and successfully decrypted a book). The problem is that the Nook Android app seems to forbid backups so you'd need a way to extract that user data. But looking at @momijizukamori there seems to be a way to do that. Or did you use an emulator and extracted the files from that? I used my actual Android phone and modded the app to allow backups.

**Comment by noDRM** [comment]
Next version of the plugin will support importing keys from the Nook Windows Store app, and also from the extracted Android app data. 

If you want to test without waiting for the next release, you can click on the "Actions" tab at the top, then click on the newest workflow run, then on the "plugin" artifact at the bottom and you'll be able to download a ZIP file with the newest plugin packaged from the latest commit.

**Comment by momijizukamori** [comment]
> The problem is that the Nook Android app seems to forbid backups so you'd need a way to extract that user data (either a rooted device or a patched APK). But looking at @momijizukamori's response, there seems to be a way to do that. Or did you use an emulator and extracted the files from that? I used my actual Android phone and modded the app to allow backups.

I have a rooted Android phone (Pixel 3a) so I just used a file browser app to copy the files out of the system data dir and into a userspace folder so I could email them to myself.

**Comment by junefish** [comment]
These instructions just worked perfectly for me! https://github.com/noDRM/DeDRM_tools/discussions/9
